File: gfg/2024_8_august/c++/24_0_-_1_knapsack_problem.cpp
// TIME COMPEXITY
// O(n * w)

// SPACE COMPEXITY
// O(n * w)

// CODE

class Solution {
  public:
    // Function to return max value that can be put in knapsack of capacity W.
    
    int solve(int ind, int W, vector<int>& wt, vector<int>& val, vector<vector<int> > &dp){
        if(ind == 0){
            if(wt[0] <= W) return val[0];
            else return 0;
        }
        
        if(dp[ind][W] != -1) return dp[ind][W];
        
        int include = 0;
        if(wt[ind] <= W){
            include = val[ind] + solve(ind-1,W-wt[ind],wt,val,dp);
        }
        int exclude = solve(ind-1,W,wt,val,dp);
        
        return dp[ind][W] = max(include,exclude);
    }
    
    int knapSack(int W, vector<int>& wt, vector<int>& val) {
        int n = wt.size();
        vector<vector<int> > dp(n, vector<int> (W+1,-1));
        return solve(n-1,W,wt,val,dp);
    }
};

File: gfg/2024_8_august/python/24_01_knapsack_problem.py
# CODE

class Solution:
    def knapSack(self, w, wt, val):
        n = len(wt)
        dp = [[-1 for _ in range(w + 1)] for _ in range(n + 1)]
        for i in range(n + 1):
            dp[i][0] = 0
        for i in range(w + 1):
            dp[0][i] = 0
        for i in range(1, n + 1):
            for j in range(1, w + 1):
                if wt[i - 1] <= j:
                    include = val[i - 1] + dp[i - 1][j - wt[i - 1]]
                    exclude = dp[i - 1][j]
                    dp[i][j] = max(include, exclude)
                else:
                    dp[i][j] = dp[i - 1][j]
        return dp[n][w]